base/int.jl

```diff
@@ -514,3 +514,140 @@ checked_mul(x::Int128, y::Int128) = x * y
+        signedmax::typeof(d) = signedmin - 1
+        allones = (zero(d) - 1) % T
+
+        nc::typeof(d) = allones - rem(convert(T, allones - d), d)
+        p = 8*sizeof(d) - 1                    # initialize p
+        q1::typeof(d), r1::typeof(d) = divrem(signedmin, nc)
+        q2::typeof(d), r2::typeof(d) = divrem(signedmax, d)
+        while true
+            p += 1
+            if r1 >= convert(T, nc - r1)
+                q1 = q1 + q1 + T(1)     # update q1
+                r1 = r1 + r1 - nc       # update r1
+            else
+                q1 = q1 + q1            # update q1
+                r1 = r1 + r1            # update r1
+            end
+            if convert(T, r2 + T(1)) >= convert(T, d - r2)
+                add |= q2 >= signedmax
+                q2 = q2 + q2 + 1        # update q2
+                r2 = r2 + r2 + T(1) - d # update r2
+            else
+                add |= q2 >= signedmin
+                q2 = q2 + q2            # update q2
+                r2 = r2 + r2 + T(1)     # update r2
+            end
+            delta::typeof(d) = d - 1 - r2
+            (p < sizeof(d)*16 && (q1 < delta || (q1 == delta && r1 == 0))) || break
+        end
+        m = q2 + 1                   # resulting magic number
+        s = p - sizeof(d)*8 - add    # resulting shift
+        new(d, m % T, add, s % UInt8)
+    end
+end
+UnsignedFastDivInteger(x::Unsigned) = UnsignedFastDivInteger{typeof(x)}(x)
+
+# Special type to handle div by 1
+immutable FastDivInteger1{T} <: FastDivInteger{T} end
+
+(*)(a::FastDivInteger, b::FastDivInteger) = a.divisor*b.divisor
+(*)(a::Number, b::FastDivInteger) = a*b.divisor
+(*)(a::FastDivInteger, b::Number) = a.divisor*b
+
+# div{T}(a::Integer, b::SignedFastDivInteger{T})   = div(convert(T, a), b)
+# div{T}(a::Integer, b::UnsignedFastDivInteger{T}) = div(convert(T, a), b)
+# rem{T}(a::Integer, b::FastDivInteger{T})         = rem(convert(T, a), b)
+# divrem{T}(a::Integer, b::FastDivInteger{T})      = divrem(convert(T, a), b)
+
+function div{T}(a::T, b::SignedFastDivInteger{T})
+    x = ((widen(a)*b.multiplier) >>> sizeof(a)*8) % T
+    x += (a*b.addmul) % T
+    (signbit(x) + (x >> b.shift)) % T
+end
+function div{T}(a::T, b::UnsignedFastDivInteger{T})
+    x = ((widen(a)*b.multiplier) >>> sizeof(a)*8) % T
+    x = ifelse(b.add, convert(T, convert(T, (convert(T, a - x) >>> 1)) + x), x)
+    x >>> b.shift
+end
+div{T}(a, ::FastDivInteger1{T}) = convert(T, a)
+
+rem{T}(a::T, b::FastDivInteger{T}) =
+    a - div(a, b)*b.divisor
+rem{T}(a, ::FastDivInteger1{T}) = zero(T)
+
+function divrem{T}(a::T, b::FastDivInteger{T})
+    d = div(a, b)
+    (d, a - d*b.divisor)
+end
+divrem{T}(a, ::FastDivInteger1{T}) = (convert(T, a), zero(T))
+
+# Type unstable!
+call(::Type{FastDivInteger}, x::Signed) = x == 1 ? FastDivInteger1{typeof(x)}() : SignedFastDivInteger(x)
+call(::Type{FastDivInteger}, x::Unsigned) = x == 1 ? FastDivInteger1{typeof(x)}() : UnsignedFastDivInteger(x)
```

Comment:
I was hoping to avoid an extra `ifelse` and do it by clever arithmetic, if possible. It's already slower than I'd like, but perhaps the flip side is that an extra `ifelse` won't make it much worse. Within the world of ReshapedArrays the type instability does not add any extra challenges, since currently their creation is type unstable (see follow-up comment below).

But if we expect this to be used in other contexts, then perhaps the type instability is something we need to fix, even if it slows down the usage.